sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-calculation/e-cert-calculation-process.ts

```diff
@@ -0,0 +1,140 @@
+          disbursementsPerStudent[uniqueStudentId],
+          log,
+        ),
+      uniqueStudentIds,
+    );
+  }
+
+  /**
+   * Group all disbursements per student id.
+   * @param eCertDisbursements disbursements to be grouped.
+   * @returns grouped disbursements.
+   */
+  private getGroupedDisbursementsPerStudent(
+    eCertDisbursements: EligibleECertDisbursement[],
+  ): GroupedDisbursementPerStudent {
+    return eCertDisbursements.reduce(
+      (
+        group: { [studentId: number]: EligibleECertDisbursement[] },
+        disbursement,
+      ) => {
+        const studentId = disbursement.studentId;
+        if (!group[studentId]) {
+          group[studentId] = [];
+        }
+        group[studentId].push(disbursement);
+        return group;
+      },
+      {},
+    );
+  }
+
+  /**
+   * Execute all the disbursements for a single student sequentially.
+   * @param eCertDisbursements all disbursements for a single student.
+   * @param parentLog cumulative process log.
+   */
+  private async calculateDisbursementsForStudent(
+    eCertDisbursements: EligibleECertDisbursement[],
+    parentLog: ProcessSummary,
+  ): Promise<void> {
+    const steps = this.calculationSteps();
+    // Ensures that disbursements for the same student will be processed
+    // sequentially since one disbursement may affect student account data like
+    // overawards balance and maximums where a second disbursement should take
+    // those updated data into consideration.
+    for (const eCertDisbursement of eCertDisbursements) {
+      const disbursementLog = new ProcessSummary();
+      parentLog.children(disbursementLog);
+      disbursementLog.info(
+        `Processing disbursement id ${eCertDisbursement.disbursement.id} scheduled for ${eCertDisbursement.disbursement.disbursementDate}.`,
+      );
+      try {
+        await this.dataSource.transaction(async (entityManager) => {
+          let stepNumber = 1;
+          // Execute all steps sequentially. The order of execution is important since the
+          // disbursement data is potentially changed along the steps till it is persisted.
+          for (const step of steps) {
+            disbursementLog.info(
+              `Executing step ${stepNumber++} out of ${steps.length}.`,
+            );
+            const shouldProceed = await step.executeStep(
+              eCertDisbursement,
+              entityManager,
+              disbursementLog,
+            );
+            if (!shouldProceed) {
+              disbursementLog.info(
+                "The step determined that the calculation should be interrupted. This disbursement will not be part of the next e-Cert generation.",
+              );
+              break;
+            }
+          }
+        });
+      } catch (error: unknown) {
+        disbursementLog.error(
+          "Unexpected error while processing disbursement. Student disbursements for this student will no longer be calculated.",
```

Comment:
It will be nice to have the disbursement id in the log.

Reply:
The logs do not need to be standalone information when they are grouped. In this case, it was committed because the previous log entry to `disbursementLog` does exactly it.

Reply:
Got it. Makes sense.